sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs

```diff
@@ -0,0 +1,456 @@
+enum Groups {
+    BLS12381Scalar = 0,
+    BLS12381G1 = 1,
+    BLS12381G2 = 2,
+    BLS12381GT = 3,
+}
+
+impl Groups {
+    fn from_u8(value: u8) -> Option<Self> {
+        match value {
+            0 => Some(Groups::BLS12381Scalar),
+            1 => Some(Groups::BLS12381G1),
+            2 => Some(Groups::BLS12381G2),
+            3 => Some(Groups::BLS12381GT),
+            _ => None,
+        }
+    }
+}
+
+fn parse<G: ToFromByteArray<S>, const S: usize>(e: &[u8]) -> FastCryptoResult<G> {
+    G::from_byte_array(e.try_into().map_err(|_| FastCryptoError::InvalidInput)?)
+}
+
+// Binary operations with 2 different types.
+fn binary_op_diff<
+    G1: ToFromByteArray<S1>,
+    G2: ToFromByteArray<S2>,
+    const S1: usize,
+    const S2: usize,
+>(
+    op: impl Fn(G1, G2) -> FastCryptoResult<G2>,
+    a1: &[u8],
+    a2: &[u8],
+) -> FastCryptoResult<Vec<u8>> {
+    let e1 = parse::<G1, S1>(a1)?;
+    let e2 = parse::<G2, S2>(a2)?;
+    let result = op(e1, e2)?;
+    Ok(result.to_byte_array().to_vec())
+}
+
+// Binary operations with the same type.
+fn binary_op<G: ToFromByteArray<S>, const S: usize>(
+    op: impl Fn(G, G) -> FastCryptoResult<G>,
+    a1: &[u8],
+    a2: &[u8],
+) -> FastCryptoResult<Vec<u8>> {
+    binary_op_diff::<G, G, S, S>(op, a1, a2)
+}
+
+// TODO: Since in many cases more than one group operation will be performed in a single
+// transaction, it might be worth caching the affine representation of the group elements and use
+// them to save conversions.
+
+pub fn internal_validate(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 2);
+
+    // TODO: charge fees
+    let cost = context.gas_used();
+    if !is_supported(context) {
+        return Ok(NativeResult::err(cost, NOT_SUPPORTED_ERROR));
+    }
+
+    let bytes_ref = pop_arg!(args, VectorRef);
+    let bytes = bytes_ref.as_bytes_ref();
+    let group_type = pop_arg!(args, u8);
+
+    let result = match Groups::from_u8(group_type) {
+        Some(Groups::BLS12381Scalar) => {
+            parse::<bls::Scalar, { bls::Scalar::BYTE_LENGTH }>(&bytes).is_ok()
+        }
+        Some(Groups::BLS12381G1) => {
+            parse::<bls::G1Element, { bls::G1Element::BYTE_LENGTH }>(&bytes).is_ok()
+        }
+        Some(Groups::BLS12381G2) => {
+            parse::<bls::G2Element, { bls::G2Element::BYTE_LENGTH }>(&bytes).is_ok()
+        }
+        Some(Groups::BLS12381GT) => {
+            parse::<bls::GTElement, { bls::GTElement::BYTE_LENGTH }>(&bytes).is_ok()
+        }
+        _ => false,
+    };
+
+    Ok(NativeResult::ok(cost, smallvec![Value::bool(result)]))
+}
+
+pub fn internal_add(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 3);
+
+    // TODO: charge fees
```

Comment:
adding gas fee is pretty vital and I am not sure I see what we are doing here.
Can you please explain?